File: Forasoft/JavaVersion/app/src/main/java/com/pustovit/forajava/view/adapter/AlbumDiffUtil.java
package com.pustovit.forajava.view.adapter;

import androidx.recyclerview.widget.DiffUtil;
import com.pustovit.forajava.repository.model.Album;
import java.util.List;

/**
 * Created by Pustovit Vladimir on 25.01.2020.
 * vovapust1989@gmail.com
 */

 class AlbumDiffUtil extends DiffUtil.Callback{
    private List<Album> oldList;
    private List<Album> newList;

    public AlbumDiffUtil(List<Album> oldList, List<Album> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList == null ? 0 : oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList == null ? 0 : newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).getCollectionId().equals(newList.get(newItemPosition).getCollectionId());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).getCollectionId().equals(newList.get(newItemPosition).getCollectionId());
    }
}

File: Forasoft/KotlinVersion/app/src/main/java/com/pustovit/kotlinversion/view/adapter/AlbumDiffUtil.kt
package com.pustovit.kotlinversion.view.adapter

import androidx.recyclerview.widget.DiffUtil
import com.pustovit.kotlinversion.repository.model.Album

/**
 * Created by Pustovit Vladimir on 29.01.2020.
 * vovapust1989@gmail.com
 */

internal class AlbumDiffUtil(
    oldList: List<Album>,
    newList: List<Album>
) :
    DiffUtil.Callback() {
    private val oldList: List<Album>?
    private val newList: List<Album>?
    override fun getOldListSize(): Int {
        return oldList?.size ?: 0
    }

    override fun getNewListSize(): Int {
        return newList?.size ?: 0
    }

    override fun areItemsTheSame(
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        return oldList!![oldItemPosition].collectionId == newList!![newItemPosition].collectionId
    }

    override fun areContentsTheSame(
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        return oldList!![oldItemPosition].collectionId == newList!![newItemPosition].collectionId
    }

    init {
        this.oldList = oldList
        this.newList = newList
    }
}